pkg/symbol/elfutils/debuginfofile.go

```diff
@@ -130,9 +157,124 @@ func (f *debugInfoFile) SourceLines(addr uint64) ([]profile.LocationLine, error)
 		})
 	}
 
+	name, ok := tr.Entry.Val(dwarf.AttrName).(string)
+	if !ok {
+		name = ""
+	}
+
+	// address correspond line must in last position
+	file, line := findLineInfo(f.lineEntries[cu.Offset], addr)
+	lines = append(lines, profile.LocationLine{
+		Line: line,
+		Function: f.demangler.Demangle(&pb.Function{
+			Name:     name,
+			Filename: file,
+		}),
+	})
+
+	moveLocationLinesForwardOneStep(lines)
 	return lines, nil
 }
 
+// moveLinesForwardOneStep move each LocationLine's line and file
+// forward one step to get right line number and file.
+//
+// Example program:
+// ```
+// 0 package main
+// 1
+// 2 func main() {
+// 3      Top()
+// 4 }
+// 5 func Top() {
+// 6       // pc
+// 7  }
+// ```
+// Use dwarf to tran pc to lines.
+// The function top is inline subroutine,
+// It's line would be [main.Top main.go 3],
+// (the line number is where the top be called as inlined).
+// And the pc's dwarf.LineEntry
+// is [main.main main.go 6]. (actual line number)
+// So the result is [{main.Top main.go 3}, {main.main main.go 6}].
+//
+// But the pprof format expected [{main.Top main.go 6}, {main.main main.go 3}].
+// It wants function line number and file not where it called, but itself.
+func moveLocationLinesForwardOneStep(locationLines []profile.LocationLine) {
```

Comment:
this function doesn't make sense to me, dwarf entries are hierarchical, so we might have something like this.

Just examples, but this should demonstrate it:

```
dwarf.Entry{AttrName: "main.main", range: [0x1, 0x6]}
  dwarf.Entry{AttrName: "main.Top", AttrCallLine: 6, range: [0x3, 0x4]}
```

The `AttrCallLine` denotes the line number of the `main.main` parent.

Then the line table is separate and only notes the beginnings of address ranges, file names, and line numbers (not function names):

```
0x1: main.go line number: 4
0x2: main.go line number: 5
0x3: top.go line number: 1
0x4: top.go line number: 2
0x5: main.go line number: 7
0x6: main.go line number: 8
```

Combining these we can now say that the full stacks are the following for the program counter ranges:

```
[0x1, 0x2): [{main.main main.go 4}]
[0x2, 0x3): [{main.main main.go 5}]
[0x3, 0x4): [{main.Top main.go 1}, {main.main main.go 6}]
[0x4, 0x5): [{main.Top main.go 2}, {main.main main.go 6}]
[0x5, 0x6): [{main.main main.go 7}]
[0x6, 0x7): [{main.main main.go 8}]
```

Reply:
> this function doesn't make sense to me, dwarf entries are hierarchical, so we might have something like this.
> 
> Just examples, but this should demonstrate it:
> 
> ```
> dwarf.Entry{AttrName: "main.main", range: [0x1, 0x6]}
>   dwarf.Entry{AttrName: "main.Top", AttrCallLine: 6, range: [0x3, 0x4]}
> ```
> 
> The `AttrCallLine` denotes the line number of the `main.main` parent.
> 
> Then the line table is separate and only notes the beginnings of address ranges, file names, and line numbers (not function names):
> 
> ```
> 0x1: main.go line number: 4
> 0x2: main.go line number: 5
> 0x3: top.go line number: 1
> 0x4: top.go line number: 2
> 0x5: main.go line number: 7
> 0x6: main.go line number: 8
> ```
> 
> Combining these we can now say that the full stacks are the following for the program counter ranges:
> 
> ```
> [0x1, 0x2): [{main.main main.go 4}]
> [0x2, 0x3): [{main.main main.go 5}]
> [0x3, 0x4): [{main.Top main.go 1}, {main.main main.go 6}]
> [0x4, 0x5): [{main.Top main.go 2}, {main.main main.go 6}]
> [0x5, 0x6): [{main.main main.go 7}]
> [0x6, 0x7): [{main.main main.go 8}]
> ```

in your example, main.Top have wrong AttrCallLine, it should be 3, but not 6.
because AttrCallLine is where main.Top being called.

and only with function moveLocationLinesForwardOneStep,
the output consistent with addr2line and net/http/pprof,
which i think is correct.

Reply:
`moveLocationLinesForwardOneStep` may make the input "correct", but this means we're not generating the frames correctly in the first place. There should be no need to stitch things together like this.

Reply:
> `moveLocationLinesForwardOneStep` may make the input "correct", but this means we're not generating the frames correctly in the first place. There should be no need to stitch things together like this.
i need correct funcname, line number and file to view source code.

And i do not actually understand the meaning of 'correct frame',
addr2line writen by linux, any user would use it to tran address to funcname,
if user could see consistent output, 
it is difficult to explain why to their.